### DISCUSSION TITLE
Ability to set Xcode build flags

### DISCUSSION THREAD

**Discussion by elihwyma** [SwiftPackageIndex/SwiftPackageIndex-Server#3470] [Ideas]
Currently I don't see anyway to set Xcode build flags in a .spi.yml configuration file. This means that packages that include C++ code, and would need the `SWIFT_OBJC_INTEROP_MODE = objcxx;` Xcode variable will not build from Xcode. An example package is [PugiSwift](https://swiftpackageindex.com/elihwyma/PugiSwift)

**Comment by finestructure** [comment]
How do you currently pass this parameter to your build command? I would expect it to be either via build settings in `Package.swift` or via a setting in you scheme in your `xcodeproj` or `xcconfig` file.

**Reply by elihwyma** [reply]
This is a bit of an annoyance with Swift, you have to specify the flag in both your xcodeproj and Package.swift file. Currently it will build fine with `swift build` but if you do it through Xcode it will fail unless you add the flag. Does the build system use a .xcconfig file already?

**Reply by elihwyma** [reply]
Ha I found the bug, apologies 🥲

**Comment by elihwyma** [comment]
Reopening this as I found the exact issue. The build command setting the OTHER_SWIFT_FLAGS causes the Swift flags in the Package.swift to be ignored, so the custom flag required (C++ Interop) is ignored causing the compile to fail.

**Reply by finestructure** [reply]
Could you point me to a failing build to get a better understanding what the setup looks like and what our build command is?

**Reply by elihwyma** [reply]
Sure, [this build](https://swiftpackageindex.com/builds/88B987AB-005C-4ABF-9507-738EA9FDBA32) that uses Xcode for macOS fails whereas [this build](https://swiftpackageindex.com/builds/EE2B3695-B3BB-41B0-9177-59683A8DDE98) succeeds.

**Reply by finestructure** [reply]
For later reference (because the builds will eventually disappear) this is about the package [PugiSwift](https://swiftpackageindex.com/elihwyma/PugiSwift).

Thanks for those links, @elihwyma . I see and can reproduce the issue now. Unfortunately, I don't think that's something that's easy to fix. I've run a test to try and preserve the setting while also adding the Swift 6 flags we're injecting but that test was unsuccessful.

Note that the injection of `OTHER_SWIFT_FLAGS` for `xcodebuild` builds is something we will only be doing while we are running the [Ready for Swift 6](https://swiftpackageindex.com/ready-for-swift-6) project tracking Swift 6 adoption. While we don't have an exact timeline in mind when we'll stop running these "special" builds, it's not something we plan on doing permanently.

That said, I wanted to note that the reported build matrix on your package page would not actually change at all if this `xcodebuild` passed. Here's what it looks like right now:

<img width="537" alt="Screenshot 2024-11-03 at 15 37 36" src="https://github.com/user-attachments/assets/3626d264-4b07-4908-9e4e-6e8d5cc22e35">

You'll notice that we show both Swift 6 and macOS compatibility and that's due to the way we aggregate the individual results up. The passing SwiftPM build ensures that your package gets credit for this combination and the failing `xcodebuild` doesn't enter the picture.

Does that work for your package in the interim?

**Reply by elihwyma** [reply]
The only part of the matrix I would like to have "changed" is the platform support. It should be usable on every platform. The Swift 6.0 only is because I use some of the new syntax.
